build_zng.rs

```diff
@@ -1,58 +1,691 @@
-use std::env;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    env, fs,
+    hash::{Hash as _, Hasher as _},
+    io::Write as _,
+    path::{Path, PathBuf},
+};
+
+fn append(cfg: &mut cc::Build, root: Option<&str>, files: impl IntoIterator<Item = &'static str>) {
+    let root = root.unwrap_or("");
+    cfg.files(
+        files
+            .into_iter()
+            .map(|fname| format!("src/zlib-ng/{root}{fname}.c")),
+    );
+}
+
+/// Replicate the behavior of cmake/make/configure of stripping out the
+/// @ZLIB_SYMBOL_PREFIX@ since we don't want or need it
+fn strip_symbol_prefix(input: &Path, output: &Path, mut on_line: impl FnMut(&str)) {
+    let contents = fs::read_to_string(input).unwrap();
+    let mut h = fs::File::create(output).expect("failed to create zlib include");
+
+    use std::io::IoSlice;
+    let mut write = |bufs: &[IoSlice]| {
+        // write_all_vectored is unstable
+        for buf in bufs {
+            h.write_all(&buf).unwrap();
+        }
+    };
+
+    for line in contents.lines() {
+        if let Some((begin, end)) = line.split_once("@ZLIB_SYMBOL_PREFIX@") {
+            write(&[
+                IoSlice::new(begin.as_bytes()),
+                IoSlice::new(end.as_bytes()),
+                IoSlice::new(b"\n"),
+            ]);
+        } else {
+            write(&[IoSlice::new(line.as_bytes()), IoSlice::new(b"\n")]);
+        }
+
+        on_line(line);
+    }
+}
+
+#[derive(Default)]
+struct AppendState {
+    flags: BTreeSet<&'static str>,
+    defines: BTreeSet<&'static str>,
+    files: BTreeSet<&'static str>,
+}
+
+#[derive(Debug, Hash, Copy, Clone)]
+struct TargetFeature {
+    check: &'static str,
+    msvc_flags: &'static [&'static str],
+    flags: &'static [&'static str],
+    defines: &'static [&'static str],
+    files: &'static [&'static str],
+}
+
+struct Ctx<'ctx> {
+    cfg: &'ctx mut cc::Build,
+    root: Option<&'ctx str>,
+    enabled: Option<&'ctx BTreeSet<&'ctx str>>,
+    msvc: bool,
+    append: AppendState,
+    cache: Option<BTreeMap<u64, bool>>,
+}
+
+impl<'ctx> Ctx<'ctx> {
+    fn enable(&mut self, tf: TargetFeature) -> bool {
+        if let Some(enabled) = self.enabled {
```

Comment:
You can use a match here to avoid having two ifs:

```suggestion
        match self.enabled {
            Some(enabled) if !enabled.contains(tf.check) => false,
            _ => {
                self.push(tf);
                true
            }
        }
```